src/test/java/integration/EnrollmentTests.java

```diff
@@ -0,0 +1,153 @@
+        assertEquals(student.getId(),updatedStudent.getId());
+
+        // Verify that student's and course's value of coins changed
+        assertNotEquals(student.getCoins(), updatedStudent.getCoins());
+        var updatedCourse = courseService.getById(course.getId());
+        assertNotEquals(course.getCoinsPaid(), updatedCourse.getCoinsPaid());
+    }
+
+    @Test
+    void enrollStudentEnrollmentDenied() throws Exception {
+        // Create groups of courses and students, make sure that the course is private
+        course.getSettings().setIsPublic(false);
+        course = createCourse(course);
+        student = createStudent(student);
+
+        // Perform enrollment and check response status
+        mockMvc.perform(post("/courses/{id}/students/{studentId}", course.getId(), student.getId()))
+                .andExpect(status().isInternalServerError());
+    }
+
+    @Test
+    void enrollStudentInsufficientFunds() throws Exception {
+        /* Create course and student, make sure that the course is public,
+           and the price is higher than the number of student coins
+         */
+        course.getSettings().setIsPublic(true);
+        course = createCourse(course);
+
+        student.setCoins(course.getPrice().subtract(BigDecimal.ONE));
+        student = createStudent(student);
+
+        // Perform enrollment and check response status
+        mockMvc.perform(post("/courses/{id}/students/{studentId}", course.getId(), student.getId()))
+                .andExpect(status().isInternalServerError());
+    }
+
+    @Test
+    void unenrollStudentSuccess() throws Exception {
+        /* Create course and student, make sure that the course is public,
+           and the number of student coins is not lower than the price
+         */
+        course.getSettings().setIsPublic(true);
+        course = createCourse(course);
+        student.setCoins(course.getPrice());
+        student = createStudent(student);
+
+        // Perform enrollment
+        mockMvc.perform(post("/courses/{id}/students/{studentId}", course.getId(), student.getId()))
+                .andExpect(status().isOk());
+        // Perform unenrolment
+        mockMvc.perform(delete("/courses/{id}/students/{studentId}", course.getId(), student.getId()))
+                .andExpect(status().isNoContent());
+
+        // Verify that course doesn't contain students
+        var courseStudents = studentService.getAllByCourseId(course.getId());
+        assertEquals(0, courseStudents.size());
+    }
+
+    private Course createCourse(Course course) {
+        course.setId(null);
+        course.getSettings().setId(null);
+        return courseService.create(course);
+    }
+
+    private Student createStudent(Student student) {
+        student.setId(null);
+        student.setEnrollments(null);
+        student.setEmail(String.format("%s@gmail.com", UUID.randomUUID()));
+        return studentService.create(student);
+    }
```

Comment:
Like in code with business logic where we are splitting logic by different classes, it's also better to create some helper classes, e.g. `CourseBuilder, StudentBuilder` in order to split test classes and data builders